### PR TITLE
fix(Redirect): Better handle wrong redirect header in a response

### LINKED ISSUE #1386: Uncatchable error when server redirects to invalid URL
**Reproduction**

1. Request to a valid URL to which the server responds with an invalid Location header
2. Server crashes despite `await fetch()` being wrapped in `try { } / catch () { }`

```
Uncaught TypeError [ERR_INVALID_URL]: Invalid URL: //super:invalid:url%/
      at onParseError (internal/url.js:279:9)
      at new URL (internal/url.js:355:5)
      at ClientRequest.<anonymous> (file:///home/tasinet/code/node-fetch/src/index.js:133:52)
      at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:647:27)
      at HTTPParser.parserOnHeadersComplete (_http_common.js:126:17)
      at Socket.socketOnData (_http_client.js:515:22)
      at addChunk (internal/streams/readable.js:290:12)
      at readableAddChunk (internal/streams/readable.js:265:9)
      at Socket.Readable.push (internal/streams/readable.js:204:10)
      at TCP.onStreamRead (internal/stream_base_commons.js:188:23)
```

An example URL with this behaviour is `http://vardakeio.gr`:
```
curl -v http://vardakeio.gr 2>&1 | grep Location
< Location: https://vardakeio.gr%/
```

**Expected behavior**

Invalid URL should be caught and an appropriate error returned

**Your Environment**

| software         | version
| ---------------- | -------
| node-fetch       | 3.1.0
| node             | 14.17.6
| npm              | 6.14.15
| Operating System | Ubuntu 20.04.2 LTS

**Additional context**

This is because the location header is passed to `new URL()` without a try/catch block around it:

index.js:133
```
// HTTP fetch step 5.3
const locationURL = location === null ? null : new URL(location, request.url);
```

Raising a PR to address this.

### PR DESCRIPTION
**Pull request by tasinet**
## Purpose
Fixes crash when a Location header point to an invalid URL

## Changes
Just wrapping `new URL()` in a try/catch and returning a sensible (?) error message

## Additional information
Test included

___

- [x] I updated ./docs/CHANGELOG.md with a link to this PR or Issue
- [x] I added unit test(s)

___

- Fixes #1386

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -130,7 +130,15 @@ export default async function fetch(url, options_) {
 				const location = headers.get('Location');
 
 				// HTTP fetch step 5.3
-				const locationURL = location === null ? null : new URL(location, request.url);
+				let locationURL;
+				try {
+					locationURL = location === null ? null : new URL(location, request.url);
+				} catch (error) {
+					// error can only be invalid URL
+					reject(new FetchError(`uri requested responds with an invalid redirect URL: ${location}`, 'invalid-redirect'));
```
Comment (LinusU) [inline src/index.js:138]: Have we looked at what error browsers gives back here? 🤔

Comment (LinusU) [reply src/index.js:138]: Chrome:

Screenshot 2021-11-18 at 16 37 18

Safari:

Screenshot 2021-11-18 at 16 37 43

Right, browsers doesn't give any more info on errors 😅

**Review by LinusU — APPROVED** [review]
LGTM 👍

File: src/index.js (hunk 1 above)
Comment (jimmywarting) [reply src/index.js:138]: Note of caution, Had some errors in the past where we don't return... 

```diff
- reject(...)
+ return reject(...)
```

Comment (jimmywarting) [reply src/index.js:138]: Thinking if we should throw a `TypeError` here as well to match what the browser gives back... but obvious with a better error message.

`new TypeError('Failed to fetch; Redirect url is not valid, Got: ' + location)`

Comment (LinusU) [reply src/index.js:138]: In that case I think it's true for every `FetchError`, no? 🤔

Comment (jimmywarting) [reply src/index.js:138]: > In that case I think it's true for every `FetchError`, no? 🤔

pretty much 😅

Comment (tasinet) [reply src/index.js:138]: From what I can tell, in node-fetch TypeErrors are mostly "programmer error" and FetchErrors are "remote/response error" 

Some TypeErrors:

node-fetch cannot load ${url}. URL scheme "${parsedURL.protocol.replace(/:$/, '')}" is not supported.');
Redirect option '${request.redirect}' is not a valid value of RequestRedirect'));
body used already for: ${data.url}');
Invalid referrerPolicy: ${referrerPolicy}');
Invalid referrerPolicy: ${policy}');
${parsedURL} is an url with embedded credentails.');
Request with GET/HEAD method cannot have body');
Expected signal to be an instanceof AbortSignal or EventTarget');
Header name must be a valid HTTP token [${name}]');
Invalid character in header content ["${name}"]');
Header pairs must be iterable');
Each header pair must be an iterable object');
Each header pair must be a name/value tuple');
Failed to construct \'Headers\': The provided value is not of type \'(sequence<sequence<ByteString>> or record<ByteString, ByteString>)');
Failed to fetch');
no or bad content-type header, no multipart boundary');

Some FetchErrors:

**Invalid response body while trying to fetch ${data.url}: ${error.message}**
Invalid response body while trying to fetch ${this.url}: ${error_.message}
content size at ${data.url} over limit: ${data.size}
Could not create Buffer from response body for ${data.url}: ${error.message}
Premature close of server response while trying to fetch ${data.url}

Whereas in the browser there only seem to be [AbortError and TypeErrors](https://developer.mozilla.org/en-US/docs/Web/API/fetch#exceptions)

File: src/index.js [hunk 2]
```diff
@@ -130,7 +130,15 @@ export default async function fetch(url, options_) {
 				const location = headers.get('Location');
 
 				// HTTP fetch step 5.3
-				const locationURL = location === null ? null : new URL(location, request.url);
+				let locationURL;
+				try {
+					locationURL = location === null ? null : new URL(location, request.url);
+				} catch (_error) {
```
Comment (LinusU) [inline src/index.js:136]: The catch binding is optional on all versions of Node.js that we support:

```suggestion
				} catch {
```

File: src/index.js (hunk 1 above)
Comment (LinusU) [reply src/index.js:138]: I agree, but I think that we should fix that in a separate pull request? I think that this one is ready for merging as soon as the linting error is fixed?

**Comment by jimmywarting** [comment]
Just thought of something... and i tested this in chrome...

if you have `redirect: 'manual'` then it shouldn't follow the redirect header and try to fetch it... then you may think that it should not throw any error for having invalid headers but it dose...

`fetch(/redirect/301/invalid', { redirect: 'manual' })`

so i tested it in firefox, and safari also

```js
// FF OK
fetch('http://localhost:58555/redirect/301/invalid', { redirect: 'manual' }).then(console.log, console.error)
// Safari OK
fetch('http://localhost:58555/redirect/301/invalid', { redirect: 'manual' }).then(console.log, console.error)
// Chrome Throws
fetch('http://localhost:58555/redirect/301/invalid', { redirect: 'manual' }).then(console.log, console.error)
```

We should have a test case for both.
would you mind
- Adding unit test for when `redirect` is set to `manual` and dose not throw
- Resolve the conflicts

Should also report this to Chrome / WPT

**Comment by jimmywarting** [comment]
Reported it to chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1271920

**Review by jimmywarting — CHANGES_REQUESTED** [review]
need a redirect=manual test that don't throw

**Comment by tasinet** [comment]
Do you need anything else from my end here?

**Review by LinusU — APPROVED** [review]
LGTM

**Review by jimmywarting — APPROVED** [review]

**Pull request merged**